### ISSUE TITLE
On parsing midi files, add option to retrieve absolute note on and note off times in ticks

### ISSUE THREAD

**Issue by lgmventura** [JacquesLucke/animation_nodes#1885]
I noticed, from using the node _Read Midi File_ and related notes and also by following the discussion on the [issue](https://github.com/JacquesLucke/animation_nodes/issues/1195) about creating these nodes, that the parser is converting the times _note on_ and _note off_ into seconds before outputting it. This means, it is dividing the absolute time in _ticks_ by the midi TPQ (**T**icks **P**er **Q**uarter note) so it gets the time in quarter notes. Then it runs through the list of tempos (quarter notes per minute) considering their starting times to calculate a given time in seconds.

### I would suggest adding following outputs:
- Node _Read MIDI File_:
  - List of tempos – each element would be a tempo (beats per minute or per second) and its starting time in **ticks** (int)
  - Ticks Per Quarter note (TPQ), which is a single integer for the MIDI file
- Node _MIDI Note Info_:
  - Note on in ticks (int)
  - Note off in ticks (int)
  - (rename current _note on_ and _note off_ to _note on in seconds_ and _note off in seconds_ or something like that)

### OR alternatively
- Node _Read MIDI File_:
  - List of tempos – each element would be a tempo (beats per minute or per second) and its starting time in **quarter notes** (float)
- Node _MIDI Note Info_:
  - Note on in unit quarter notes (float, already divided by TPQ)
  - Note off in unit quarter notes (float, already divided by TPQ)
  - (rename current _note on_ and _note off_ to _note on in seconds_ and _note off in seconds_ or something like that)

This would make it possible to create more complex animations – for example, taking the [Raw Note Information example](https://docs.animation-nodes.com/tutorials/midi/), we could use the tempos to pan the view with the speed proportional to the current tempo, without having to stretch the notes instead.

I could help you doing this, since I've already dealt with that in my [music visualization tool](https://github.com/lgmventura/MusicaVisual).

**Comment by OmarEmaraDev** [comment]
Since Animation Nodes is mostly made for artists, we have to make sure that any feature we add or data we expose would be somewhat intuitive to understand and use.

Temp events, TPQ, and musical units are rather hard to understand, so I am not sure about exposing them or how we should do that. Maybe you can share an example node tree that would use those data (On dummy data just for investigation sake) just to get a feel or what a typical node tree would look like?

**Comment by lgmventura** [comment]
@OmarEmaraDev I understand what you mean, but I don't think this would be more counter-intuitive than other stuff we already have in Animation Nodes. I think whoever already can use MIDI nodes and knows a bit of music would understand how to deal with unit of time in quarter notes (assuming my second implementation option above) without any problem. I'm in favour of having as much functionality as we can think of, since it is well documented with some examples. And it looks like this is already the idea of Animation Nodes.

The figure below illustrates how a node tree would look like – compare to the [original example](https://docs.animation-nodes.com/tutorials/midi/#raw-note-information). **New features** to be implemented are displayed in **green** and new nodes needed for this setup (compared to that example) are highlighted by the **orange** dashed line.

![Example node tree for panning notes in a speed proportional to current tempo, instead of stretching them. Based on the 2nd example from Animation Nodes Documentation (Raw Note Information)](https://user-images.githubusercontent.com/16710238/197405234-aef23022-9eda-4884-b830-7ec5febefa54.png)

Basically, the node _Read MIDI File_ would retrieve _Tempos_, which would be a structure containing a list of _tempo values_ and a _list of tempo starting positions_. Then, the new node _MIDI Tempos Info_ would get those lists from the structure. And _MIDI Note Info_ would get a Mode dropdown menu with the options _Time in seconds_ (current implementation) and _Time in quarter notes_. The TPQ calculation could be done in the background too, like it is being already done for the current implementation.

**Comment by lgmventura** [comment]
I forked the project and I'm implementing this. Then I could submit a pull request later and you could help me to revise the code.

**Comment by OmarEmaraDev** [comment]
@lgmventura  Feel free to open a pull request with your initial suggestion, but I wouldn't create a special structure for tempos, providing them as two arrays is fine as far as I can see.

**Comment by lgmventura** [comment]
@OmarEmaraDev I managed to make it work, but it was a bit more complicated than my first guess. Anyway, the result is amazing! Below you can see the first working version. The first image shows the first part: loading the midi, calling the subprogram and using the tempos to pan the screen accordingly. The second image shows the subprogram only. I'll write a detailed documentation later explaining the whole node tree.

![Node setup, not showing subprogram](https://user-images.githubusercontent.com/16710238/197834327-08314cb4-f4e5-4568-b943-dd2b9b28f47d.png)

![Subprogram only](https://user-images.githubusercontent.com/16710238/197834710-06549164-257f-4b11-b370-0bcba604d770.png)

I think it would be cleaner to pack the screen panning part into a new subprogram.

Maybe we could offer some node templates inside a frame or a subprogram that already does the whole thing and the user has the option to use those templates instead of setting up all nodes.

**Comment by lgmventura** [comment]
The code would also need some review and some small improvements would be good, like using the dropdown menu in _MIDI Note Info_ as I suggested before instead of showing all these outputs at the same time, if relevant.
